pkgs/ffigen/lib/src/code_generator/unique_namer.dart

```diff
@@ -0,0 +1,53 @@
+// Copyright (c) 2025, the Dart project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart_keywords.dart';
+
+class UniqueNamer {
+  final Set<String> _used;
+
+  /// Creates a UniqueNamer including all the [parent]'s names.
+  UniqueNamer({UniqueNamer? parent}) : _used = parent?._used.toSet() ?? {};
+
+  /// Creates a unique version of [name] and adds it to the set of used names.
+  String makeUnique(String name) {
+    if (name.isEmpty) {
+      // For example, nested structures/unions may not have a name.
+      name = 'unnamed';
+    }
+
+    // If the name is a keyword, append a '$'. Note that this extra '$' is
```

Comment:
Could be done in a separate issue/PR:

One thing I do for JNIgen that you might want to adopt is to be less strict with "keywords". Some of them can be used in certain contexts, also if we `import dart:core as foo`, then the symbols there such as `int` (although this particular one is also a keyword in other languages so it won't exist) can be used for names.

For example some names such as `Function` cannot be used as a class name but can be used as a field or a method name: 
https://github.com/dart-lang/native/blob/main/pkgs/jnigen/lib/src/bindings/renamer.dart#L28-L84

(Good thing I send you this because I noticed I mistakenly have `factory` as `'factory': _Allowed.fields | _Allowed.fields` instead of `'factory': _Allowed.methods | _Allowed.fields`)